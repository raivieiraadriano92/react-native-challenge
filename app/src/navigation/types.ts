import type { FunctionComponent } from "react";

import {
  CompositeScreenProps,
  NavigatorScreenParams
} from "@react-navigation/native";
import { Story } from "src/services/api/types";

import type { BottomTabScreenProps } from "@react-navigation/bottom-tabs";
import type { NativeStackScreenProps } from "@react-navigation/native-stack";

export type RootStackParamList = {
  Article: Pick<Story, "story_url">;
  Tabs: undefined | NavigatorScreenParams<TabParamList>;
};

export type RootStackScreenProps<T extends keyof RootStackParamList> =
  NativeStackScreenProps<RootStackParamList, T>;

export type RootStackScreen<RouteName extends keyof RootStackParamList> =
  FunctionComponent<RootStackScreenProps<RouteName>>;

export type TabParamList = {
  Articles: undefined;
  Settings: undefined;
};

export type TabScreenProps<T extends keyof TabParamList> = CompositeScreenProps<
  BottomTabScreenProps<TabParamList, T>,
  RootStackScreenProps<keyof RootStackParamList>
>;

export type TabScreen<RouteName extends keyof TabParamList> = FunctionComponent<
  TabScreenProps<RouteName>
>;

declare global {
  namespace ReactNavigation {
    interface RootParamList extends RootStackParamList {}
  }
}
