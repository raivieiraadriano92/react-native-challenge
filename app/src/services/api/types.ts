export type Response = {
  hits: Story[];
};

export type Story = {
  author: string;
  objectID: string;
  story_title: string;
  story_url: string;
  created_at: string;
  isDeleted?: boolean;
};
