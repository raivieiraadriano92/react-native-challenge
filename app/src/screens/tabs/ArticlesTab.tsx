import { View } from "react-native";
import { TabScreen } from "src/navigation/types";

export const ArticlesTab: TabScreen<"Articles"> = () => {
  return <View />;
};
