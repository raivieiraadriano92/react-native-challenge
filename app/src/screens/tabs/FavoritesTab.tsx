import { FlatList } from "react-native";
import { StoryRow } from "src/components/StoryRow";
import { TabScreen } from "src/navigation/types";
import { useStoriesStore } from "src/store/storiesStore";

export const FavoritesTab: TabScreen<"Favorites"> = ({ navigation }) => {
  const storiesStore = useStoriesStore();

  const favorites = storiesStore.list.filter(
    (story) => storiesStore.favorites.includes(story.objectID),
    []
  );

  return (
    <FlatList
      data={favorites}
      keyExtractor={(item) => item.objectID}
      renderItem={({ item }) => (
        <StoryRow
          isFavorite={storiesStore.favorites.includes(item.objectID)}
          onPress={() =>
            navigation.navigate("Article", { story_url: item.story_url })
          }
          onPressFavorite={() => storiesStore.toggleFavorite(item.objectID)}
          story={item}
        />
      )}
      showsVerticalScrollIndicator={false}
    />
  );
};
