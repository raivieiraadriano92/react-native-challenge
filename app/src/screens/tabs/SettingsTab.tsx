import { ScrollView, View } from "react-native";

import { SettingsRow } from "src/components/SettingsRow";
import { SettingsSection } from "src/components/SettingsSection";
import { TabScreen } from "src/navigation/types";
import { useGeneralStore } from "src/store/generalStore";
import { useStoriesStore } from "src/store/storiesStore";

export const SettingsTab: TabScreen<"Settings"> = () => {
  const storiesStore = useStoriesStore();

  const generalStore = useGeneralStore();

  return (
    <ScrollView showsVerticalScrollIndicator={false}>
      <View className="flex-1 p-6 p space-y-6">
        <SettingsSection
          description="Tailor your experience! Select your interests to receive personalized articles. Choose Android, iOS, or both to stay in the loop with the content that matters most to you."
          title="Interests"
        >
          <SettingsRow
            isActive={storiesStore.preferences.android}
            onValueChange={storiesStore.toggleAndroid}
            title="Receive articles for Android"
          />
          <SettingsRow
            isActive={storiesStore.preferences.ios}
            onValueChange={storiesStore.toggleIOS}
            title="Receive articles for iOS"
          />
        </SettingsSection>
        <View className="bg-neutral-100 h-[1px]" />
        <SettingsSection
          description="Stay updated with the latest tips, tricks, and news for your Android or iOS device! Enable notifications to never miss an article that helps you get the most out of your smartphone."
          title="Notifications"
        >
          <SettingsRow
            isActive={generalStore.notificationPreferences.android}
            onValueChange={generalStore.toggleAndroid}
            title="Receive articles for Android"
          />
          <SettingsRow
            isActive={generalStore.notificationPreferences.ios}
            onValueChange={generalStore.toggleIOS}
            title="Receive articles for iOS"
          />
        </SettingsSection>
      </View>
    </ScrollView>
  );
};
