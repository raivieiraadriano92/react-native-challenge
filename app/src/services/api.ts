import axios from "axios";

const api = axios.create({
  baseURL: "https://hn.algolia.com/api/v1"
});

export const fetchNews = (query: string) =>
  api.get(`/search_by_date?query=${query}`);
