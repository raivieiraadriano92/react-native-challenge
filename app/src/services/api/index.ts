import axios from "axios";

import { Response } from "./types";

const api = axios.create({
  baseURL: "https://hn.algolia.com/api/v1"
});

export const fetchStories = (query: string) =>
  api.get<Response>(`/search_by_date?query=${query}`);
