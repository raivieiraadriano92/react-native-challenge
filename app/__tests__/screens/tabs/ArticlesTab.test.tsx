import { fireEvent, render, screen } from "@testing-library/react-native";
import { ArticlesTab } from "src/screens/tabs/ArticlesTab";
import { Story } from "src/services/api/types";

const mockNavigationProps = {
  setOptions: jest.fn(),
  navigate: jest.fn()
};

const mockToggleDeleted = jest.fn();

const mockFetch = jest.fn();

const mockToggleFavorite = jest.fn();

const mockStory: Story = {
  author: "toomuchtodo",
  objectID: "41195352",
  story_title: "China's Real Economic Crisis",
  story_url: "https://www.foreignaffairs.com/china/chinas-real-economic-crisis",
  created_at: "2024-08-08T19:37:08Z",
  created_at_i: 1723145828,
  isDeleted: false
};

const mockDeletedStory: Story = {
  ...mockStory,
  story_title: "Deleted Story",
  objectID: "41195222",
  isDeleted: true
};

jest.mock("src/store/storiesStore", () => ({
  useStoriesStore: () => ({
    favorites: [],
    list: [mockStory, mockDeletedStory],
    fetch: mockFetch,
    toggleDeleted: mockToggleDeleted,
    toggleFavorite: mockToggleFavorite
  })
}));

afterEach(() => {
  mockFetch.mockClear();
});

describe("<ArticlesTab />", () => {
  test("should not render deleted articles", () => {
    render(<ArticlesTab navigation={mockNavigationProps} />);

    expect(() => screen.getByText(mockDeletedStory.story_title)).toThrow();

    expect(screen.getByText(mockStory.story_title)).toBeTruthy();
  });

  test("should fetch articles at first launch", () => {
    render(<ArticlesTab navigation={mockNavigationProps} />);

    expect(mockFetch).toHaveBeenCalledTimes(1);
  });

  test("should navigate to ArticleScreen", () => {
    render(<ArticlesTab navigation={mockNavigationProps} />);

    fireEvent.press(screen.getByTestId(`story-${mockStory.objectID}`));

    expect(mockNavigationProps.navigate).toHaveBeenCalledWith("Article", {
      story_url: mockStory.story_url
    });
  });

  test("should call toggleFavorite from useStoriesStore", () => {
    render(<ArticlesTab navigation={mockNavigationProps} />);

    fireEvent.press(screen.getByTestId("toogleFavoriteButton"));

    expect(mockToggleFavorite).toHaveBeenCalledTimes(1);
  });

  test("should call toggleDeleted from useStoriesStore", () => {
    render(<ArticlesTab navigation={mockNavigationProps} />);

    fireEvent.press(screen.getByTestId("deleteButton"));

    expect(mockToggleDeleted).toHaveBeenCalledTimes(1);
  });

  it("Snapshot", () => {
    const tree = render(
      <ArticlesTab navigation={mockNavigationProps} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
