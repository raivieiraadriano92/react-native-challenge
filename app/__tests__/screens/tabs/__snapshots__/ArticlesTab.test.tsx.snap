// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticlesTab /> Snapshot 1`] = `
[
  <View
    style={
      [
        {
          "paddingBottom": 24,
          "paddingLeft": 24,
          "paddingRight": 24,
          "paddingTop": 24,
        },
      ]
    }
    testID="isFetchingIndicator"
  >
    <ActivityIndicator />
  </View>,
  <RCTScrollView
    data={
      [
        {
          "author": "toomuchtodo",
          "created_at": "2024-08-08T19:37:08Z",
          "created_at_i": 1723145828,
          "isDeleted": false,
          "objectID": "41195352",
          "story_title": "China's Real Economic Crisis",
          "story_url": "https://www.foreignaffairs.com/china/chinas-real-economic-crisis",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <RCTRefreshControl />
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          collapsable={false}
          handlerTag={13}
          handlerType="PanGestureHandler"
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onLayout={[Function]}
          style={
            {
              "overflow": "hidden",
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "bottom": 0,
                "flexDirection": "row-reverse",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": [
                  {
                    "translateX": -10000,
                  },
                ],
              }
            }
          >
            <View
              style={
                [
                  {
                    "backgroundColor": "#e5e5e5",
                  },
                  {
                    "flexBasis": "0%",
                    "flexGrow": 1,
                    "flexShrink": 1,
                  },
                  {
                    "flexDirection": "row",
                  },
                  {
                    "alignItems": "center",
                  },
                ]
              }
            >
              <View
                style={
                  [
                    {
                      "flexBasis": "0%",
                      "flexGrow": 1,
                      "flexShrink": 1,
                    },
                    {
                      "paddingBottom": 24,
                      "paddingLeft": 24,
                      "paddingRight": 24,
                      "paddingTop": 24,
                    },
                  ]
                }
              >
                <Text
                  style={
                    [
                      {
                        "fontWeight": "700",
                      },
                      {
                        "color": "#737373",
                      },
                      {
                        "fontSize": 18,
                        "lineHeight": 28,
                      },
                    ]
                  }
                >
                  Are you sure?
                </Text>
              </View>
              <View
                collapsable={false}
                style={
                  {
                    "opacity": 0,
                  }
                }
              >
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "#ef4444",
                      "flexBasis": "0%",
                      "flexGrow": 1,
                      "flexShrink": 1,
                      "height": "100%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "paddingBottom": 24,
                      "paddingLeft": 24,
                      "paddingRight": 24,
                      "paddingTop": 24,
                    }
                  }
                  testID="deleteButton"
                >
                  <Text
                    style={
                      {
                        "color": "#fcfcfc",
                        "fontWeight": "bold",
                        "padding": 3,
                      }
                    }
                  >
                    Delete
                  </Text>
                </View>
              </View>
            </View>
            <View
              onLayout={[Function]}
            />
          </View>
          <View
            collapsable={false}
            handlerTag={14}
            handlerType="TapGestureHandler"
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            pointerEvents="auto"
            style={
              {
                "transform": [
                  {
                    "translateX": -0,
                  },
                ],
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "backgroundColor": "#fff",
                  "borderBottomColor": "#f5f5f5",
                  "borderBottomWidth": 1,
                  "borderLeftColor": "#f5f5f5",
                  "borderRightColor": "#f5f5f5",
                  "borderTopColor": "#f5f5f5",
                  "flexDirection": "row",
                  "opacity": 1,
                  "paddingBottom": 24,
                  "paddingLeft": 24,
                  "paddingRight": 24,
                  "paddingTop": 24,
                }
              }
              testID="story-41195352"
            >
              <View
                style={
                  [
                    {
                      "flexBasis": "0%",
                      "flexGrow": 1,
                      "flexShrink": 1,
                    },
                  ]
                }
              >
                <Text
                  style={
                    [
                      {
                        "fontWeight": "500",
                      },
                      {
                        "fontSize": 16,
                        "lineHeight": 24,
                      },
                      {
                        "color": "#FF6600",
                      },
                    ]
                  }
                >
                  China's Real Economic Crisis
                </Text>
                <Text
                  style={
                    [
                      {
                        "fontSize": 14,
                        "lineHeight": 20,
                      },
                      {
                        "color": "#a3a3a3",
                      },
                    ]
                  }
                >
                  toomuchtodo - 1 day
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "marginLeft": 24,
                    "opacity": 1,
                  }
                }
                testID="toogleFavoriteButton"
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#fbbf24",
                        "fontSize": 24,
                      },
                      undefined,
                      {
                        "fontFamily": "FontAwesome",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;
